File: func-pipes/static/js/power2/nodes/meter.js
/*
  nodes/meter.js — Instrument / Power Meter
  ─────────────────────────────────────────────────────────────────────────────
  A read-only instrument: measures and displays V / A / W, then passes
  the signal through to downstream nodes unchanged.

  Extended state
  ──────────────
  volts  number  — measured voltage
  amps   number  — measured current
  watts  number  — computed power (V × A)

  States: 'off' | 'on'
*/

class Meter extends NodeBase {

    static type  = 'meter'
    static label = 'Meter'
    static group = 'Instrument'

    static catalog = [
        { key: 'meter', label: 'Power Meter' },
    ]

    static defaults(id, preset = {}) {
        return {
            ...super.defaults(id, preset),
            volts: 0,
            amps:  0,
            watts: 0,
        }
    }

    static configFields() {
        return [...super.configFields()]   // no extra config beyond label
    }

    /**
     * Measure the inbound signal and pass it through unchanged. Readings are
     * throttled and emitted as 'meter:reading' events when V or A changes.
     * @param {Object}     panel
     * @param {Object|null} signal — upstream { v, a } or null
     * @param {PowerGraph} graph
     */
    static apply(panel, signal, graph) {
        if (!signal || signal.v <= 0) {
            const prev  = panel.state
            panel.state = 'off'
            panel.volts = 0
            panel.amps  = 0
            panel.watts = 0
            if (prev !== 'off') Meter.dispatch(panel, 'state:change', { from: prev, to: 'off' })
            graph.emit(panel, null)
            return
        }

        const prev  = panel.state
        const volts = +signal.v.toFixed(1)
        const amps  = +signal.a.toFixed(2)
        const watts = +(signal.v * signal.a).toFixed(0)

        panel.state = 'on'
        panel.volts = volts
        panel.amps  = amps
        panel.watts = watts

        if (prev !== 'on') Meter.dispatch(panel, 'state:change', { from: prev, to: 'on' })

        if (volts !== panel._lastVolts || amps !== panel._lastAmps) {
            panel._lastVolts = volts
            panel._lastAmps  = amps
            Meter.throttle(panel, 'meter:reading', { volts, amps, watts })
        }

        graph.emit(panel, signal)   // transparent pass-through
    }

    /**
     * Reset all readings to zero, clear internal de-dupe tracking, and
     * propagate to NodeBase.reset().
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static reset(panel, graph) {
        panel.volts        = 0
        panel.amps         = 0
        panel.watts        = 0
        panel._lastVolts   = null
        panel._lastAmps    = null
        panel.powerSources = {}
        Meter.dispatch(panel, 'meter:reset', {})
        super.reset(panel, graph)
    }
}

NodeRegistry.register(Meter)

File: power_graph/src/power_graph/nodes/meter.py
"""
nodes/meter.py
──────────────────────────────────────────────────────────────────────────────
Meter — transparent pass-through that samples voltage, amps and watts.

States: 'off' | 'on'

The meter does not modify the signal in any way; it simply records readings
and emits them as throttled events so that dashboards can display live values.
"""

from typing import Dict, List
from power_graph.node_base import NodeBase, Signal
from power_graph.node_registry import NodeRegistry


class Meter(NodeBase):
    """Inline measuring instrument — transparent pass-through."""

    type  = 'meter'
    label = 'Meter'
    group = 'Instrumentation'
    dispatch_delay = 120

    catalog = [
        {'key': 'meter-v',  'label': 'Voltmeter'},
        {'key': 'meter-a',  'label': 'Ammeter'},
        {'key': 'meter-vaw','label': 'Multi-Meter'},
    ]

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        base = super().defaults(node_id, preset)
        base.update({
            'reading_volts': 0.0,
            'reading_amps':  0.0,
            'reading_watts': 0.0,
            'state':         'off',
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields()]

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        if signal is None:
            if panel['state'] != 'off':
                cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'off'})
            panel.update({'state': 'off', 'reading_volts': 0.0,
                          'reading_amps': 0.0, 'reading_watts': 0.0})
            graph.emit(panel, None)
            return

        v = signal.get('v', 0.0)
        a = signal.get('a', 0.0)
        w = round(v * a, 2)

        changed = (
            panel.get('reading_volts') != v or
            panel.get('reading_amps')  != a
        )

        panel['reading_volts'] = v
        panel['reading_amps']  = a
        panel['reading_watts'] = w

        if panel['state'] != 'on':
            cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'on'})
        panel['state'] = 'on'

        if changed:
            cls.throttle(panel, 'meter:reading', {'v': v, 'a': a, 'w': w})

        # Pass signal through unmodified
        graph.emit(panel, signal)

    @classmethod
    def reset(cls, panel: Dict, graph):
        prev = panel['state']
        panel.update({'state': 'off', 'reading_volts': 0.0,
                      'reading_amps': 0.0, 'reading_watts': 0.0})
        cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})
        graph.emit(panel, None)


NodeRegistry.register(Meter)
